components/modal/index.js

```diff
@@ -0,0 +1,176 @@
+	 */
+	prepareDOM() {
+		if ( ! parentElement ) {
+			parentElement = document.createElement( 'div' );
+			document.body.appendChild( parentElement );
+		}
+		this.node = document.createElement( 'div' );
+		parentElement.appendChild( this.node );
+	}
+
+	/**
+	 * Removes the specific mounting point for this modal from the DOM.
+	 */
+	cleanDOM() {
+		parentElement.removeChild( this.node );
+	}
+
+	/**
+	 * Appends the modal's node to the DOM, so the portal can render the
+	 * modal in it. Also calls the openFirstModal when this is the first modal to be
+	 * opened.
+	 */
+	componentDidMount() {
+		openModalCount++;
+
+		if ( openModalCount === 1 ) {
+			this.openFirstModal();
+		}
+	}
+
+	/**
+	 * Removes the modal's node from the DOM. Also calls closeLastModal when this is
+	 * the last modal to be closed.
+	 */
+	componentWillUnmount() {
+		openModalCount--;
+
+		if ( openModalCount === 0 ) {
+			this.closeLastModal();
+		}
+
+		this.cleanDOM();
+	}
+
+	/**
+	 * Prepares the DOM for this modal and any additional modal to be mounted.
+	 *
+	 * It appends an additional div to the body for the modals to be rendered in,
+	 * it hides any other elements from screen-readers and adds an additional class
+	 * to the body to prevent scrolling while the modal is open.
+	 */
+	openFirstModal() {
+		ariaHelper.hideApp( parentElement );
+		document.body.classList.add( this.props.bodyOpenClassName );
+	}
+
+	/**
+	 * Cleans up the DOM after the last modal is closed and makes the app available
+	 * for screen-readers again.
+	 */
+	closeLastModal() {
+		document.body.classList.remove( this.props.bodyOpenClassName );
+		ariaHelper.showApp();
+	}
+
+	/**
+	 * Renders the modal.
+	 *
+	 * @return {WPElement} The modal element.
+	 */
+	render() {
+		const {
+			overlayClassName,
+			className,
+			onRequestClose,
+			title,
+			icon,
+			closeButtonLabel,
+			children,
+			aria,
+			...otherProps
+		} = this.props;
+
+		return createPortal(
+			<div className={ classnames(
+				'components-modal__screen-overlay',
+				overlayClassName
+			) }>
+				<ModalFrame
+					className={ classnames(
+						'components-modal__frame',
+						className
+					) }
+					onRequestClose={ onRequestClose }
+					aria={ {
+						labelledby: this.headingId,
+						describedby: aria.describedby,
+					} }
+					{ ...otherProps } >
+					<ModalHeader
+						closeLabel={ closeButtonLabel }
+						onClose={ onRequestClose }
+						title={ title }
+						headingId={ this.headingId }
+						icon={ icon } />
+					<div
+						className={ 'components-modal__content' }>
+						{ children }
+					</div>
+				</ModalFrame>
+			</div>,
+			this.node
+		);
+	}
+}
+
+Modal.defaultProps = {
+	bodyOpenClassName: 'modal-open',
+	role: 'dialog',
+	title: null,
+	onRequestClose: noop,
+	focusOnMount: true,
+	shouldCloseOnEsc: true,
+	shouldCloseOnClickOutside: true,
+	style: {
+		content: null,
+		overlay: null,
+	},
+	/* accessibility */
+	aria: {
+		labelledby: 'modal-heading',
```

Comment:
We're being inconsistent in the implementation about whether we expect there to be one or multiple modals present in the page at a given time. See https://github.com/WordPress/gutenberg/pull/6261/files#r196023169 . If we expect more than one, we can't use a static ID, and in general this is why a static ID should be discouraged for components in general.

Reply:
I think this is handled internally with `this.headingId = uniqueId( props.aria.labelledby )` now, so maybe the default prop doesn't make much sense now? However, please see my last comment.

Reply:
Well at first I had removed the default value from `defaultProps`, and my logic was the following:
```
this.headingId = uniqueId( this.props.aria.labelledby || 'modal-heading' );
```
But this seemed a cleaner way to approach it.